common/cuda_hip/factorization/elimination_forest_kernels.cpp

```diff
@@ -0,0 +1,384 @@
+    // reach the last node and we don't need to sort since the COO format
+    // already sorts by row index.
+    const auto policy = thrust_policy(exec);
+    const auto nnz = exec->copy_val_to_host(row_ptrs + size);
+    // convert edges to COO representation
+    // the edge list is sorted, since we only consider edges where row > col,
+    // and the row array (= weights) is sorted coming from row_ptrs
+    array<IndexType> row_array{exec, static_cast<size_type>(nnz)};
+    const auto rows = row_array.get_data();
+    components::convert_ptrs_to_idxs(exec, row_ptrs, size, rows);
+    // we assume the matrix is symmetric, so we can remove every second edge
+    // also round up the worklist size for equal cache alignment between fields
+    const auto worklist_size =
+        ceildiv(nnz, config::warp_size * 2) * config::warp_size;
+    // create 2 worklists consisting of (start, end, edge_id)
+    array<IndexType> worklist{exec, static_cast<size_type>(worklist_size * 6)};
+    auto wl1_source = worklist.get_data();
+    auto wl1_target = wl1_source + worklist_size;
+    auto wl1_edge_id = wl1_target + worklist_size;
+    auto wl2_source = wl1_source + 3 * worklist_size;
+    auto wl2_target = wl1_target + 3 * worklist_size;
+    auto wl2_edge_id = wl1_edge_id + 3 * worklist_size;
+    // atomic counters for worklists and output edge list
+    array<IndexType> counters{exec, 3};
+    auto wl1_counter = counters.get_data();
+    auto wl2_counter = wl1_counter + 1;
+    auto output_counter = wl2_counter + 1;
+    components::fill_array(exec, counters.get_data(), counters.get_size(),
+                           IndexType{});
+    // helpers for interacting with worklists
+    const auto clear_wl1 = [&] {
+        IndexType value{};
+        exec->copy_from(exec->get_master(), 1, &value, wl1_counter);
+    };
+    const auto clear_wl2 = [&] {
+        IndexType value{};
+        exec->copy_from(exec->get_master(), 1, &value, wl2_counter);
+    };
+    const auto get_wl1_size = [&] {
+        return exec->copy_val_to_host(wl1_counter);
+    };
+    const auto swap_wl1_wl2 = [&] {
+        std::swap(wl1_source, wl2_source);
+        std::swap(wl1_target, wl2_target);
+        std::swap(wl1_edge_id, wl2_edge_id);
+        std::swap(wl1_counter, wl2_counter);
+    };
+    // initialize every node to a singleton set
+    array<IndexType> parent_array{exec, size};
+    const auto parents = parent_array.get_data();
+    components::fill_seq_array(exec, parents, size);
+    // array storing the minimal edge adjacent to each node
+    array<IndexType> min_edge_array{exec, size};
+    const auto min_edges = min_edge_array.get_data();
+    constexpr auto min_edge_sentinel = std::numeric_limits<IndexType>::max();
+    components::fill_array(exec, min_edges, size, min_edge_sentinel);
+    // output row array, to be used in conjunction with out_cols in COO storage
+    array<IndexType> out_row_array{exec, size};
```

Comment:
does the size can be match?
in reference/omp, it use vector such that can be allocate more storage when assigning.
Ginkgo array does not have this ability.

Reply:
The size may be anywhere from 0 to size - 1, depending on how many subtrees the elimination forest consists of, this is just an upper bound (it could also be size - 1).